devito/finite_differences/differentiable.py

```diff
@@ -579,47 +599,92 @@ def __init_finalize__(self, *args, **kwargs):
         assert isinstance(d, StencilDimension) and d.symbolic_size == len(weights)
         assert isinstance(weights, (list, tuple, np.ndarray))
 
-        kwargs['scope'] = 'static'
+        try:
+            self._spacings = set().union(*[i.find(Spacing) for i in weights])
+        except AttributeError:
+            self._spacing = set()
+
+        kwargs['scope'] = 'constant'
 
         super().__init_finalize__(*args, **kwargs)
 
+    def __eq__(self, other):
+        return (isinstance(other, Weights) and
+                self.dimension is other.dimension and
+                self.name == other.name and
+                self.indices == other.indices and
+                self.weights == other.weights)
+
+    __hash__ = sympy.Basic.__hash__
+
+    def _hashable_content(self):
+        return super()._hashable_content() + (self.name,) + tuple(self.weights)
+
     @property
     def dimension(self):
         return self.dimensions[0]
 
+    @property
+    def spacings(self):
+        return self._spacings
+
     weights = Array.initvalue
 
 
 class IndexDerivative(IndexSum):
 
-    def __new__(cls, expr, dimensions, **kwargs):
-        if not (expr.is_Mul and len(expr.args) == 2):
-            raise ValueError("Expect expr*weights, got `%s` instead" % str(expr))
-        _, weights = expr.args
-        if not isinstance(weights, Weights):
-            # All of the SymPy versions we support end up placing the Weights
-            # array here, so if something changes we'll get an alarm through
-            # this exception
-            raise ValueError("Couldn't find weights array")
+    __rargs__ = ('expr', 'mapper')
+
+    def __new__(cls, expr, mapper, **kwargs):
+        dimensions = as_tuple(mapper.values())
+
+        # Detect the Weights among the arguments
+        weightss = []
+        for a in expr.args:
+            try:
+                f = a.function
+            except AttributeError:
+                continue
+            if isinstance(f, Weights):
+                weightss.append(a)
+
+        # Sanity check
+        if not (expr.is_Mul and len(weightss) == 1):
+            raise ValueError("Expect `expr*weights`, got `%s` instead" % str(expr))
+        weights = weightss.pop()
 
         obj = super().__new__(cls, expr, dimensions)
         obj._weights = weights
+        obj._mapper = frozendict(mapper)
 
         return obj
 
+    def _hashable_content(self):
+        return super()._hashable_content() + (self.mapper,)
+
     @property
     def weights(self):
         return self._weights
 
+    @property
+    def mapper(self):
+        return self._mapper
+
+    @property
+    def total_order(self):
+        iderivs = self.expr.find(IndexDerivative)
+        return 1 + max([i.total_order for i in iderivs], default=0)
```

Comment:
self is an IndexDerivative :)

Reply:
IndexDerivative don't have an order? If an `IndexDerivative` is always 1 it's gonna lead to inconsistency with the actual order of the derivative.

Reply:
If you have an IndexDerivative (say self), the total_order will definitely be `1 + X`. `1` is for self; `X` depends on the arguments. This still looks correct to me , it's also used in two tests (admittedly in a very simple way)

Reply:
>1 is for self;

But why is it one it could be any order derivative it's the representation of the stencil expression that corresponds to a derivative with `deriv_order` order not just 1

Reply:
Ah, yes, I see what you mean now, sorry!

I don't even have the deriv_order information at this point anymore. Perhaps I should attach it when IndexDerivative is created inside finite_differences.py. There it's definitely available!

Reply:
I've just realized that what I need for the heuristics I have in place in PRO is the concept of depth, not the total_order (I'm not that sophisticated yet...). So I'm renaming it as "depth", which is what the current implementation represents.